File: WU/internal/cli/args.go
package cli

import "strings"

type ArgSet struct {
	values map[string][]string
	flags  map[string]bool
}

func ParseArgs(argv []string) *ArgSet {
	m := &ArgSet{
		values: map[string][]string{},
		flags:  map[string]bool{},
	}

	isFlag := func(a string) bool {
		switch a {
		case "--select-all", "--dry-run", "--schedule-go-live",
			"--auto-install-os-upgrade", "--no-auto-install-os-upgrade",
			"--auto-install-applicable", "--no-auto-install-applicable",
			"--is-disclosure-restricted", "--publish-to-windows10s",
			"--is-reboot-required", "--is-co-engineered",
			"--is-for-unreleased-hardware", "--has-ui-software",
			"--no-ui", "--no-filter":
			return true
		default:
			return false
		}
	}

	addValue := func(k, v string) { m.values[k] = append(m.values[k], v) }
	addFlag := func(k string) { m.flags[k] = true }

	for i := 0; i < len(argv); i++ {
		a := argv[i]
		if !strings.HasPrefix(a, "--") {
			continue
		}
		if isFlag(a) {
			addFlag(a)
			continue
		}
		if i+1 >= len(argv) || strings.HasPrefix(argv[i+1], "--") {
			addValue(a, "")
			continue
		}

		if a == "--visible-to-accounts" || a == "--affected-oems" || a == "--chids" {
			for i+1 < len(argv) && !strings.HasPrefix(argv[i+1], "--") {
				i++
				addValue(a, argv[i])
			}
		} else {
			i++
			addValue(a, argv[i])
		}
	}

	return m
}

func (m *ArgSet) HasFlag(key string) bool { return m.flags[key] }

func (m *ArgSet) GetSingle(key string) string {
	if v := m.values[key]; len(v) > 0 {
		return v[0]
	}
	return ""
}

func (m *ArgSet) GetMany(key string) []string {
	if v := m.values[key]; len(v) > 0 {
		return v
	}
	return []string{}
}

File: WU-npm/src/cli/args.ts
const FLAGS = new Set([
  '--select-all', '--dry-run', '--schedule-go-live',
  '--auto-install-os-upgrade', '--no-auto-install-os-upgrade',
  '--auto-install-applicable', '--no-auto-install-applicable',
  '--is-disclosure-restricted', '--publish-to-windows10s',
  '--is-reboot-required', '--is-co-engineered',
  '--is-for-unreleased-hardware', '--has-ui-software',
  '--no-ui', '--no-filter',
])

const MULTI = new Set(['--visible-to-accounts', '--affected-oems', '--chids'])

export interface ArgSet {
  hasFlag(k: string): boolean
  getSingle(k: string): string
  getMany(k: string): string[]
}

export function parseArgs(argv: string[]): ArgSet {
  const values = new Map<string, string[]>()
  const flags = new Set<string>()
  const add = (k: string, v: string) => {
    const arr = values.get(k) ?? []
    arr.push(v)
    values.set(k, arr)
  }

  for (let i = 0; i < argv.length; i++) {
    const a = argv[i]
    if (!a.startsWith('--')) continue
    if (FLAGS.has(a)) {
      flags.add(a)
      continue
    }
    if (i + 1 >= argv.length || argv[i + 1].startsWith('--')) {
      add(a, '')
      continue
    }
    if (MULTI.has(a)) {
      while (i + 1 < argv.length && !argv[i + 1].startsWith('--')) {
        i++
        add(a, argv[i])
      }
    } else {
      i++
      add(a, argv[i])
    }
  }

  return {
    hasFlag: (k) => flags.has(k),
    getSingle: (k) => values.get(k)?.[0] ?? '',
    getMany: (k) => values.get(k) ?? [],
  }
}
